jitsi-utils-kotlin/src/main/kotlin/org/jitsi/utils/concurrent/Executors.kt

```diff
@@ -0,0 +1,120 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.utils.concurrent
+
+import org.jitsi.utils.logging2.LoggerImpl
+import org.json.simple.JSONObject
+import java.time.Duration
+import java.util.concurrent.Callable
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Future
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.LongAdder
+import kotlin.reflect.KFunction0
+
+abstract class TaskServiceWrapper<T : ExecutorService>(
+    val name: String,
+    protected val delegate: T
+) {
+    private val logger = LoggerImpl(name)
+    private val numExceptions = LongAdder()
+
+    protected fun wrapTask(task: Runnable): Runnable {
+        return Runnable {
+            try {
+                task.run()
+            } catch (t: Throwable) {
+                logger.warn("Uncaught exception: $t")
+                numExceptions.increment()
+                throw t
+            }
+        }
+    }
+
+    fun getStatsJson(): JSONObject = JSONObject().apply {
+        put("executor_class", delegate::class.simpleName)
+        put("num_exceptions", numExceptions.sum())
+
+        val ex = delegate as? ThreadPoolExecutor ?: return@apply
+        put("pool_size", ex.poolSize)
+        put("active_task_count", ex.activeCount)
+        put("completed_task_count", ex.completedTaskCount)
+        put("core_pool_size", ex.corePoolSize)
+        put("maximum_pool_size", ex.maximumPoolSize)
+        put("largest_pool_size", ex.largestPoolSize)
+        put("queue_class", ex.queue.javaClass.simpleName)
+        put("pending_task_count", ex.queue.size)
+    }
+
+    fun safeShutdown(timeout: Duration) = delegate.apply {
+        shutdown()
+        if (!awaitTermination(timeout.toMillis() / 2, TimeUnit.MILLISECONDS)) {
+            shutdownNow()
+            if (!awaitTermination(timeout.toMillis() / 2, TimeUnit.MILLISECONDS)) {
+                throw ExecutorShutdownTimeoutException()
+            }
+        }
+    }
+}
+
+class ExecutorShutdownTimeoutException : Exception("Timed out trying to shutdown executor service")
+
+class SafeExecutor(
+    name: String,
+    delegate: ExecutorService
+) : TaskServiceWrapper<ExecutorService>(name, delegate) {
+
+    fun execute(task: Runnable) = delegate.execute(wrapTask(task))
```

Comment:
It looks like you mimic the `Executor` interface here, but also have `ExecutorService`-like methods (the submit-style methods).  Why not just mimic the `ExecutorService` API (which is what we use everywhere) for both types and have the `submit` calls and `unsafeSubmit` variants?

Reply:
`ExecutorService` extends `Executor` (so it inherits `void execute`), which is why I named the "safe" method `execute`.

Reply:
Sorry, I don't think I followed this.  Why expose `execute` at all, given nothing we use uses `Executor` directly, but instead `ExecutorService`/`ScheduledExecutorService`?

Reply:
To mimic the `ExecutorService` API -- the method which runs a task and doesn't return anything is named `execute`, while `submit` methods return a future.

Reply:
Ok, from what I saw we didn't use `execute` anywhere in the code, so think I was confused.  Are you thinking that we should change the places which don't use the future over to `execute` (from `submit`)? That's probably a good idea.

Reply:
Yes, that's what I did in the PRs in jmt/jvb